### PR TITLE
Fix handling of pkg init and test file via args

### PR DESCRIPTION
**Pull request by blueyed**
Ref: https://github.com/pytest-dev/pytest/issues/4344#issuecomment-441095934

TODO:

- [x] changelog

### CODE REVIEW

**Comment by blueyed** [comment]
Likely not the best fix, but appears to work for the added test at least.

File: src/_pytest/main.py [hunk 1]
```diff
@@ -592,7 +592,7 @@ def filter_(f):
                 col = self._node_cache[argpath]
             else:
                 collect_root = self._pkg_roots.get(argpath.dirname, self)
-                col = collect_root._collectfile(argpath)
+                col = collect_root._collectfile(argpath, handle_dupes=False)
```
Comment (blueyed) [inline src/_pytest/main.py]: That's wrong.. and will not handle dupes at all then.
Also the output of the test should be matched (includes `sub/test_file.py` twice).

**Comment by blueyed** [comment]
This fails with 3.9.3 also already:
```
============================= test session starts ==============================
platform linux -- Python 3.7.2, pytest-3.9.3, py-1.7.0, pluggy-0.8.1 -- /home/daniel/Vcs/pytest/.venv/bin/python
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/home/daniel/Vcs/pytest/.hypothesis/examples')
rootdir: /tmp/pytest-of-daniel/pytest-270/test_collect_pkg_init_and_file_in_args0, inifile:
plugins: forked-1.0.1, hypothesis-4.0.0
collecting ... collected 3 items

sub/test_file.py::test_file PASSED                                       [ 33%]
sub/test_file.py::test_file PASSED                                       [ 33%]
sub/test_file.py::test_file PASSED                                       [ 33%]

=========================== 3 passed in 0.01 seconds ===========================
```

With 3.6.4 it looks better:
```
sub/__init__.py::test_init PASSED                                        [ 33%]
sub/test_file.py::test_file PASSED                                       [ 66%]
sub/test_file.py::test_file PASSED                                       [ 66%]
```

But percentage is still off there, and it should not run `test_file` twice (due to not using `--keepduplicates`) I assume?

**Comment by blueyed** [comment]
Re collection the file twice: this is documented, but could be fixed/improved likely: https://github.com/pytest-dev/pytest/blame/5ca81596bbbb35da0b8b742afaf7ed2ef075ccda/doc/en/example/pythoncollection.rst#L76-L84

**Comment by blueyed** [comment]
Needs `-o python_files=*.py` to handle `__init__.py` after all..

File: src/_pytest/main.py (hunk 1 above)
Comment (blueyed) [reply src/_pytest/main.py]: Likely works better with `add_to_dupes` (https://github.com/pytest-dev/pytest/pull/4748).

**Comment by blueyed** [comment]
I think this is a good small fix for now.

File: testing/test_collection.py [hunk 2]
```diff
@@ -1157,3 +1157,20 @@ def test_collectignore_via_conftest(testdir, monkeypatch):
 
     result = testdir.runpytest()
     assert result.ret == EXIT_NOTESTSCOLLECTED
+
+
+def test_collect_pkg_init_and_file_in_args(testdir):
```
Comment (nicoddemus) [inline testing/test_collection.py]: Hmm this test is passing for me without your patch (`master`). Can you double check that this is reproducing the problem for you please?

Comment (blueyed) [reply testing/test_collection.py]: Oh, snap - thanks for catching.

The difference/fix is only without `-o python_files=*.py` then - extended the test.
Not sure if this is really expected, but better then not collection anything I guess.

Comment (nicoddemus) [reply testing/test_collection.py]: I see, thanks. I checked and the new test does fail with `master`. 👍

**Comment by blueyed** [comment]
Added changelog entry.
Let me know if it is good and I'd like to squash it.

**Comment by blueyed** [comment]
Squashed.

**Pull request merged**